### PR TITLE
Make ConvolutionalFeatureMap use linalg convolution

### PR DESCRIPTION
**Pull request by sanuj**
Will push more commits to this. All unit tests are passing. Yet to add `NATIVE` implementation to linalg convolution. #2957

### CODE REVIEW

File: src/shogun/neuralnets/ConvolutionalFeatureMap.cpp [hunk 1]
```diff
@@ -283,6 +287,32 @@ void CConvolutionalFeatureMap::convolve(
 			outputs.matrix+i*outputs.num_rows + outputs_row_offset,
 			m_output_height, m_output_width, false);
 
+#if defined(HAVE_LINALG_LIB)
```
Comment (sanuj) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: @karlnapf This is one way to take care of `m_autoencoder_position` but is inefficient. A better way would be to edit linalg convolution to tackle `m_autoencoder_position` like it's done below in the existing code. If we do the other way then I'll have to edit `VIENNACL` implementation as well.

File: src/shogun/neuralnets/ConvolutionalFeatureMap.cpp [hunk 2]
```diff
@@ -283,6 +287,32 @@ void CConvolutionalFeatureMap::convolve(
...
+		if(m_autoencoder_position == NLAP_NONE)
+		{
```
Comment (karlnapf) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: minor: we do not use ``` {} ``` for single line loops/if

File: src/shogun/neuralnets/ConvolutionalFeatureMap.cpp (hunk 1 above)
Comment (karlnapf) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: I would go for the more efficient one.

File: src/shogun/neuralnets/ConvolutionalFeatureMap.cpp [hunk 3]
```diff
@@ -309,6 +339,7 @@ void CConvolutionalFeatureMap::convolve(
 				result(res_y,res_x) = sum;
 			}
 		}
+#endif
```
Comment (karlnapf) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: So this makes our NNs dependent on linalg, right?
@lisitsyn do you object?

File: src/shogun/neuralnets/ConvolutionalFeatureMap.cpp (hunk 1 above)
Comment (yorkerlin) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: Using gist for the wiki thing?

Comment (karlnapf) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: @yorkerlin ???

Comment (yorkerlin) [inline src/shogun/neuralnets/ConvolutionalFeatureMap.cpp]: Sorry. Wrong email

**Comment by karlnapf** [comment]
This is way too old

**Pull request closed without merging**